Firebase.agent.lib.nut

```diff
@@ -641,24 +655,57 @@ class Firebase {
                 }
                 if (200 <= res.statuscode && res.statuscode < 300) {
                     onSuccess(data);
+                    _tooManyReqTimer = false;
+                } else if (res.statuscode == 28 || res.statuscode == 429) {
+                    local now = time();
+                    // Too many requests, set _tooManyReqTimer to prevent more requests to FB
+                    if (_tooManyReqTimer == false) {
+                        // This is the first 429 we have seen set a default timeout
+                        _tooManyReqTimer = now + DEFAULT_BACK_OFF_TIMEOUT_SEC;
+                    } else if (_tooManyReqTimer <= now) {
+                        // Firebase is still overwhelmed after first timeout expired, 
+                        // Let's block requests for longer to let FB recover
+                        _tooManyReqCounter += 1
+                        _tooManyReqTimer = now + (DEFAULT_BACK_OFF_TIMEOUT_SEC * _tooManyReqCounter);
+                    }
+                    // Pass error to callback
+                    onError("Error " + res.statuscode);
                 } else if (typeof data == "table" && "error" in data) {
+                    _tooManyReqTimer = false;
                     local error = data ? data.error : null;
                     onError(error);
                 } else {
+                    _tooManyReqTimer = false;
                     onError("Error " + res.statuscode);
                 }
             } catch (err) {
+                _tooManyReqTimer = false;
                 onError(err);
             }
         }
     }
 
-    // use Promise if promise libary included and callback != null
     function _processResponse(request, callback) {
-        if (_promiseIncluded && callback == null) {
-            return _createRequestPromise(request);
+        // Use Promise if promise libary included and callback != null
+        local usePromise = (_promiseIncluded && callback == null);
+
+        // Only send request if we haven't received a 429 error recently
+        if (_tooManyReqTimer == false || _tooManyReqTimer <= time()) {
+            if (usePromise) {
+                return _createRequestPromise(request);
+            } else {
+                return _sendRequest(request, callback);
+            }
         } else {
-            return _sendRequest(request, callback);
+            local error = "ERROR: Too many requests to Firebase, try request again after " + _tooManyReqTimer;
```

Comment:
Will a timestamp be printed here? Will it be more user friendly to say "in ... sec"?